web/middlewares/secure.go

```diff
@@ -0,0 +1,152 @@
+package middlewares
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/instance"
+	"github.com/labstack/echo"
+)
+
+type (
+	// XFrameOption type for the values of the X-Frame-Options header.
+	XFrameOption string
+
+	// CSPSource type are the different types of CSP headers sources definitions.
+	// Each source type defines a different acess policy.
+	CSPSource int
+
+	// SecureConfig defines the config for Secure middleware.
+	SecureConfig struct {
+		HSTSMaxAge     time.Duration
+		CSPDefaultSrc  []CSPSource
+		CSPScriptSrc   []CSPSource
+		CSPFrameSrc    []CSPSource
+		CSPConnectSrc  []CSPSource
+		CSPFontSrc     []CSPSource
+		CSPImgSrc      []CSPSource
+		CSPManifestSrc []CSPSource
+		CSPMediaSrc    []CSPSource
+		CSPObjectSrc   []CSPSource
+		CSPStyleSrc    []CSPSource
+		CSPWorkerSrc   []CSPSource
+		XFrameOptions  XFrameOption
+		XFrameAllowed  string
+	}
+)
+
+const (
+	// XFrameDeny is the DENY option of the X-Frame-Options header.
+	XFrameDeny XFrameOption = "DENY"
+	// XFrameSameOrigin is the SAMEORIGIN option of the X-Frame-Options header.
+	XFrameSameOrigin = "SAMEORIGIN"
+	// XFrameAllowFrom is the ALLOW-FROM option of the X-Frame-Options header. It
+	// should be used along with the XFrameAllowed field of SecureConfig.
+	XFrameAllowFrom = "ALLOW-FROM"
+
+	// CSPSrcSelf is the 'self' option of a CSP source.
+	CSPSrcSelf CSPSource = iota
+	// CSPSrcParent adds the parent domain as an eligible CSP source.
+	CSPSrcParent
+	// CSPSrcParentSubdomains add all the parent's subdomains as eligibles CSP
+	// sources.
+	CSPSrcParentSubdomains
+	// CSPSrcAny is the '*' option. It allows any domain as an eligible source.
+	CSPSrcAny
+)
+
+// Secure returns a Middlefunc that can be used to define all the necessary
+// secure headers. It is configurable with a SecureConfig object.
+func Secure(conf *SecureConfig) echo.MiddlewareFunc {
+	var hstsHeader string
+	if conf.HSTSMaxAge > 0 {
+		hstsHeader = fmt.Sprintf("max-age=%.f; includeSubDomains",
+			conf.HSTSMaxAge.Seconds())
+	}
+
+	var xFrameHeader string
+	switch conf.XFrameOptions {
+	case XFrameDeny:
+		xFrameHeader = string(XFrameDeny)
+	case XFrameSameOrigin:
+		xFrameHeader = string(XFrameSameOrigin)
+	case XFrameAllowFrom:
+		xFrameHeader = fmt.Sprintf("%s %s", XFrameAllowFrom, conf.XFrameAllowed)
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			hsts := true
+			if !c.IsTLS() {
```

Comment:
I'm not sure about this condition. When we have a reverse-proxy in front of cozy-stack, there will an https connexion between the client/browser and the reverse-proxy, but the connexion between the reverse-proxy and the stack can be http or https (and if the reverse-proxy is on the same server that the stack, it will be probably http, not https). The reverse-proxy can set the `X-Forwarded-Protocol` header, but the question is: does `c.IsTLS()` checks this header? From what I've read, I think it is not the case.

So, I think we should remove this condition for the moment. And add it again later if we have some issues, and testing how it behaves with a reverse-proxy.